lib/java/persistance/src/main/java/org/enso/persist/PerBufferReference.java

```diff
@@ -1,21 +1,28 @@
 package org.enso.persist;
 
+import java.io.IOException;
 import org.enso.persist.PerInputImpl.InputCache;
 import org.enso.persist.Persistance.Reference;
 
 final class PerBufferReference<T> extends Persistance.Reference<T> {
   private final Persistance<T> p;
   private final PerInputImpl.InputCache cache;
   private final int offset;
+  private Object cached;
 
-  private PerBufferReference(Persistance<T> p, PerInputImpl.InputCache buffer, int offset) {
+  private PerBufferReference(
+      Persistance<T> p, PerInputImpl.InputCache buffer, int offset, boolean allowCaching) {
     this.p = p;
     this.cache = buffer;
     this.offset = offset;
+    this.cached = allowCaching ? null : this;
   }
 
   @SuppressWarnings(value = "unchecked")
-  final <T> T readObject(Class<T> clazz) {
+  final <T> T readObject(Class<T> clazz) throws IOException {
+    if (cached != this && clazz.isInstance(cached)) {
```

Comment:
Similarly to @radeusgd, I probably don't understand how `cached` is supposed to be used/what it represents. This makes this condition hard to decipher

Reply:
It is a tri-state value as explained in 0b09978841